amm/contracts/stable_pool/token_rate.rs

```diff
@@ -0,0 +1,103 @@
+use ink::{contract_ref, env::DefaultEnvironment, primitives::AccountId};
+use scale::{Decode, Encode};
+use traits::RateProvider;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct ExternalTokenRate {
+    cached_token_rate: u128,
+    last_token_rate_update_ts: u64,
+    token_rate_contract: AccountId,
+    expiration_duration_ms: u64,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum TokenRate {
+    Constant(u128),
+    External(ExternalTokenRate),
+}
+
+impl TokenRate {
+    pub fn new_constant(rate: u128) -> Self {
+        Self::Constant(rate)
+    }
+
+    pub fn new_external(
+        current_time: u64,
+        token_rate_contract: AccountId,
+        expiration_duration_ms: u64,
+    ) -> Self {
+        Self::External(ExternalTokenRate::new(
+            current_time,
+            token_rate_contract,
+            expiration_duration_ms,
+        ))
+    }
+
+    pub fn get_rate(&self) -> u128 {
+        match self {
+            Self::Constant(rate) => *rate,
+            Self::External(external) => external.get_rate(),
+        }
+    }
+
+    pub fn update_rate(&mut self, current_time: u64) {
+        match self {
+            Self::External(external) => external.update_rate(current_time),
+            _ => {}
+        }
+    }
+
+    pub fn update_rate_no_cache(&mut self, current_time: u64) {
+        match self {
+            Self::External(external) => external.update_rate_no_cache(current_time),
+            _ => {}
+        }
+    }
+}
+
+impl ExternalTokenRate {
+    pub fn new(
+        current_time: u64,
+        token_rate_contract: AccountId,
+        expiration_duration_ms: u64,
+    ) -> Self {
+        let rate = Self::query_rate(token_rate_contract);
+        Self {
+            cached_token_rate: rate,
+            last_token_rate_update_ts: current_time,
+            token_rate_contract,
+            expiration_duration_ms,
+        }
+    }
+
+    pub fn get_rate(&self) -> u128 {
+        self.cached_token_rate
+    }
+
+    pub fn update_rate(&mut self, current_time: u64) {
+        if self.is_outdated(current_time) {
```

Comment:
It is possible via `update_rate_no_cache`, and there is `force_update_rate` message to trigger it in the contract. 

The default is though to call `update_rate` at the beginning of each contract call that uses rates. Because of that, you want to do it only if necessary, to avoid paying the fee when it's not necessary. 

Also it's worth noting that in practice updating the rate VERY often is not necessary, because it changes little by little only. Like 0.0003 daily or so. Updating once per day should be good.